File: scripts/charts/chart_spec_patches.py
"""
ChartSpec 补丁表 — 与 src/lib/chart-spec-patches.ts 对齐。
热路径重渲由 TS chart-runner 驱动；本模块供同进程复用 / 对照。
"""
from __future__ import annotations

import re
from copy import deepcopy
from typing import Any

_COL_UNIT_RE = re.compile(r"^(.+?)\s*[（(]([^）)]+)[）)]\s*$")


def _unit_from_name(name: str) -> str | None:
    m = _COL_UNIT_RE.match(name.strip())
    if not m:
        return None
    return f"{m.group(1).strip()} ({m.group(2).strip()})"


def _next_legend(current: str | None) -> str | None:
    if current in (None, "", "auto"):
        return "outer-right"
    if current == "outer-right":
        return "outer-bottom"
    return None


def apply_chart_spec_patches(
    spec: dict[str, Any],
    findings: list[dict[str, Any]],
) -> tuple[dict[str, Any], list[dict[str, Any]]]:
    next_spec = deepcopy(spec)
    patches: list[dict[str, Any]] = []
    codes = {
        str(f.get("code"))
        for f in findings
        if f.get("action") == "repair" or (f.get("action") == "block" and f.get("code") == "missing_unit")
    }
    layout = next_spec.setdefault("layout", {"legend": "auto"})
    if not isinstance(layout, dict):
        layout = {"legend": "auto"}
        next_spec["layout"] = layout

    if "label_overlap" in codes or "annotation_clipped" in codes:
        before = layout.get("xTickRotation") or 0
        if before < 35:
            layout["xTickRotation"] = 35
            patches.append({
                "code": "label_overlap" if "label_overlap" in codes else "annotation_clipped",
                "path": "layout.xTickRotation",
                "before": before,
                "after": 35,
            })

    if "legend_covers_data" in codes:
        after = _next_legend(layout.get("legend"))
        if after:
            patches.append({
                "code": "legend_covers_data",
                "path": "layout.legend",
                "before": layout.get("legend"),
                "after": after,
            })
            layout["legend"] = after

    encoding = next_spec.get("encoding") if isinstance(next_spec.get("encoding"), dict) else {}
    if "missing_unit" in codes and not encoding.get("unitless"):
        y_label = str(encoding.get("yLabel") or "")
        if not _COL_UNIT_RE.search(y_label):
            names: list[str] = []
            columns = next_spec.get("data", {}).get("columns") if isinstance(next_spec.get("data"), dict) else None
            if isinstance(columns, dict) and isinstance(columns.get("y"), list):
                names.extend(str(x) for x in columns["y"])
            for name in names:
                filled = _unit_from_name(name)
                if filled:
                    patches.append({
                        "code": "missing_unit",
                        "path": "encoding.yLabel",
                        "before": y_label,
                        "after": filled,
                    })
                    encoding["yLabel"] = filled
                    next_spec["encoding"] = encoding
                    break

    return next_spec, patches

File: src/lib/chart-spec-patches.ts
/**
 * FIG-QA-005：qaReport findings → ChartSpec 补丁（纯函数）。
 * 与 scripts/charts/chart_spec_patches.py 补丁表对齐。
 */
import type { ChartQaFinding } from "@/contracts/chart-qa";
import type { ChartLegendPlacement, ChartSpecLayout, ChartSpecV1 } from "@/contracts/chart-spec";
import { hasAxisUnit, parseChartSpec } from "@/contracts/chart-spec";

const COL_UNIT_RE = /^(.+?)\s*[（(]([^）)]+)[）)]\s*$/;

export interface ChartSpecPatch {
  code: string;
  path: string;
  before: string | number | boolean | null;
  after: string | number | boolean | null;
}

export interface ApplyChartSpecPatchesResult {
  spec: ChartSpecV1;
  patches: ChartSpecPatch[];
}

function cloneSpec(spec: ChartSpecV1): ChartSpecV1 {
  return JSON.parse(JSON.stringify(spec)) as ChartSpecV1;
}

function ensureLayout(spec: ChartSpecV1): ChartSpecLayout {
  if (!spec.layout) {
    spec.layout = { legend: "auto" };
  }
  return spec.layout;
}

function unitFromName(name: string): { base: string; unit: string } | null {
  const m = name.trim().match(COL_UNIT_RE);
  if (!m) return null;
  return { base: m[1].trim(), unit: m[2].trim() };
}

function yLabelFromColumns(spec: ChartSpecV1): string | null {
  const names = [
    ...(spec.data.columns?.y ?? []),
    ...(spec.data.datasets?.map((d) => d.label) ?? []),
  ];
  for (const name of names) {
    const parsed = unitFromName(name);
    if (parsed) return `${parsed.base} (${parsed.unit})`;
  }
  return null;
}

function nextLegend(current: ChartLegendPlacement | undefined): ChartLegendPlacement | null {
  if (!current || current === "auto") return "outer-right";
  if (current === "outer-right") return "outer-bottom";
  return null;
}

function addPatch(
  patches: ChartSpecPatch[],
  code: string,
  path: string,
  before: ChartSpecPatch["before"],
  after: ChartSpecPatch["after"],
): void {
  if (before === after) return;
  patches.push({ code, path, before, after });
}

/** 只消费 repair（以及可回填的 missing_unit block） */
export function isPatchableFinding(finding: ChartQaFinding): boolean {
  if (finding.action === "repair") return true;
  return finding.action === "block" && finding.code === "missing_unit";
}

export function applyChartSpecPatches(
  spec: ChartSpecV1,
  findings: readonly ChartQaFinding[],
): ApplyChartSpecPatchesResult {
  const next = cloneSpec(spec);
  const patches: ChartSpecPatch[] = [];
  const codes = new Set(findings.filter(isPatchableFinding).map((f) => f.code));

  if (codes.has("label_overlap") || codes.has("annotation_clipped")) {
    const layout = ensureLayout(next);
    const before = layout.xTickRotation ?? 0;
    if (before < 35) {
      layout.xTickRotation = 35;
      addPatch(patches, codes.has("label_overlap") ? "label_overlap" : "annotation_clipped", "layout.xTickRotation", before, 35);
    }
  }

  if (codes.has("legend_covers_data")) {
    const layout = ensureLayout(next);
    const after = nextLegend(layout.legend);
    if (after) {
      addPatch(patches, "legend_covers_data", "layout.legend", layout.legend, after);
      layout.legend = after;
    }
  }

  if (codes.has("missing_unit") && !next.encoding.unitless && !hasAxisUnit(next.encoding.yLabel)) {
    const filled = yLabelFromColumns(next);
    if (filled) {
      addPatch(patches, "missing_unit", "encoding.yLabel", next.encoding.yLabel, filled);
      next.encoding.yLabel = filled;
    }
  }

  if (codes.has("width_off_spec")) {
    const cols = next.journal.columns === 2 ? 2 : 1;
    addPatch(patches, "width_off_spec", "journal.columns", next.journal.columns, cols);
    next.journal.columns = cols;
  }

  const parsed = parseChartSpec(next);
  return { spec: parsed ?? spec, patches };
}
